### ISSUE TITLE
Idea: convert our episode audio files into mp4, then upload to YouTube automatically

### ISSUE THREAD

**Issue by LinuxSuRen** [opensource-f2f/episode#244]
YouTube already have [the Postcast support](https://www.youtube.com/@osf2f/podcasts). It would be great to convert our episode audio files into mp4, then upload to YouTube automatically. For example, we could find or create a CLI to do it. Then create a GitHub workflow to trigger it.

I found some potential tools could help us:

* https://github.com/tokland/youtube-upload

**Comment by hellojukay** [comment]
What is the workflow? Where can I get the mp3 files and YouTube video title information? It seems that I only need to write some short scripts to achieve it.

**Comment by wjsvec** [comment]
Will there be such a risk?
![image](https://github.com/opensource-f2f/episode/assets/39830125/fb3c1878-cfc1-4022-9bca-766c27ec4d5f)

**Comment by wey-gu** [comment]
Maybe the concern comes from running CPU-intensive tasks (i.e. [FFmpeg converting job](https://github.com/marketplace/actions/setup-ffmpeg)) with GH Action?

@yihong0618 had a lot of hacking GH action experiences to alert such concern, maybe please you could share more context?

**Comment by yihong0618** [comment]
@LinuxSuRen @wey-gu do some search, for this job I think its OK.
but what I am concern is that for a official account it still be risk, something like https://www.v2ex.com/t/874221

**Comment by wjsvec** [comment]
I will try to implement this issue. Here's a question.

The "Check if the episode exists in YouTube" step may cause problems due to anti-crawlers. Can we save a log file locally to record the uploaded videos?

**Comment by wey-gu** [comment]
> I will try to implement this issue. Here's a question.
> 
> 
> 
> The "Check if the episode exists in YouTube" step may cause problems due to anti-crawlers. Can we save a log file locally to record the uploaded videos?
> 
> 
> 
> 

That will do!

Another mitigation to avoid being treated as malicious probe would be checking via rss interface/ proxied rss:

https://www.youtube.com/feeds/videos.xml?channel_id=THE_CHANNEL_ID_HERE

While, I agree with you that in current slo, maintain local state may be much better in many perspectives.

**Comment by wjsvec** [comment]
> > I will try to implement this issue. Here's a question.
> > The "Check if the episode exists in YouTube" step may cause problems due to anti-crawlers. Can we save a log file locally to record the uploaded videos?
> 
> That will do!
> 
> Another mitigation to avoid being treated as malicious probe would be checking via rss interface/ proxied rss:
> 
> https://www.youtube.com/feeds/videos.xml?channel_id=THE_CHANNEL_ID_HERE
> 
> While, I agree with you that in current slo, maintain local state may be much better in many perspectives.

aha, what is the channel ID of osf2f? It seems that only the author can see this ID himself.

**Comment by LinuxSuRen** [comment]
> aha, what is the channel ID of osf2f? It seems that only the author can see this ID himself.

Here is the channel. https://www.youtube.com/channel/UCV7Ibg1k_aMSEcDFgJvuvdg

But please make sure that everything is ok before using it. And thanks for your effort.

**Comment by hellojukay** [comment]
我尝试去解决这个问题，但是遇到了2个问题 https://github.com/hellojukay/episode/actions

1. 在 github action 里面进行 mp3 转 mp4 速度非常慢，一个视频大概需要 40 多分钟
2. 上传 mp4 到 youtube 认证需要浏览器重定向点击认证，也可以使用 token , 但是 token 七天会过期
3. 判断 mp4 在 youtube 是否已经存在，可能比较麻烦，需要调用 api 之类的，也存在应用认证的问题。

**Comment by wey-gu** [comment]
> 2. 上传 mp4 到 youtube 认证需要浏览器重定向点击认证，也可以使用 token , 但是 token 七天会过期

It seems it was in status "Testing" causing the refresh token expiring in 7 days, for production refresh token, it's longer, someone said it's like 1 year?

> 3. 判断 mp4 在 youtube 是否已经存在，可能比较麻烦，需要调用 api 之类的，也存在应用认证的问题。

As I mentioned above, we may leverage the RSS of the channel to probe content without auth?

**Comment by LinuxSuRen** [comment]
> 在 github action 里面进行 mp3 转 mp4 速度非常慢，一个视频大概需要 40 多分钟

This is not a big problem. We can wait.

Acrrroding to @wey-gu comment, one year's token seem be acceptable.

By the way, the RSS address of YouTube seems not available. See also https://www.youtube.com/feeds/videos.xml?channel_id=UCV7Ibg1k_aMSEcDFgJvuvdg

**Comment by wey-gu** [comment]
> > 在 github action 里面进行 mp3 转 mp4 速度非常慢，一个视频大概需要 40 多分钟
> 
> This is not a big problem. We can wait.
> 
> Acrrroding to @wey-gu comment, one year's token seem be acceptable.
> 
> By the way, the RSS address of YouTube seems not available. See also https://www.youtube.com/feeds/videos.xml?channel_id=UCV7Ibg1k_aMSEcDFgJvuvdg

“https://www.youtube.com/feeds/videos.xml?channel_id=UCV7Ibg1k_aMSEcDFgJvuvdg” should be the one, maybe just accessing via browser is not allowed, I checked via feedly it's fine.

<img width="889" alt="Screenshot 2023-10-13 at 15 45 48" src="https://github.com/opensource-f2f/episode/assets/1651790/c1656d99-12d8-4cf6-a6c1-b4db836540f9">


```
<link rel="alternate" type="application/rss+xml" title="RSS" href="https://www.youtube.com/feeds/videos.xml?channel_id=UCV7Ibg1k_aMSEcDFgJvuvdg">
```

**Comment by wey-gu** [comment]
update: it seems rss for youtube is full of problems, we could think of it doesnt support rss at all :(

**Comment by LinuxSuRen** [comment]
I cannot get the correct RSS content even via the following command:

```shell
curl https://www.youtube.com/feeds/videos.xml?channel_id=UCV7Ibg1k_aMSEcDFgJvuvdg \
  -H "Content-Type: application/rss+xml" -H 'Accept: application/rss+xml'
```

**Comment by wey-gu** [comment]
> I cannot get the correct RSS content even via the following command:
> 
> ```shell
> curl https://www.youtube.com/feeds/videos.xml?channel_id=UCV7Ibg1k_aMSEcDFgJvuvdg \
>   -H "Content-Type: application/rss+xml" -H 'Accept: application/rss+xml'
> ```

Yeah, i have done the equivalent via python and encountered 404, too.
